starknet_py/net/client_models.py

```diff
@@ -1106,3 +1126,37 @@ class BlockTransactionTrace:
 
     transaction_hash: int
     trace_root: TransactionTrace
+
+
+@dataclass
+class OutsideExecution:
+    """
+    Dataclass representing an outside execution.
+    (SNIP-9)[https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-9.md]
+    """
+
+    caller: int
+    nonce: int
+    execute_after: int
+    execute_before: int
+    calls: List[Call]
+
+    def to_abi_dict(self) -> Dict:
+        """
+        Returns a dictionary that can be serialized (compiled) into calldata
+        using StructSerializer
+        """
+        return {
+            "caller": self.caller,
+            "nonce": self.nonce,
+            "execute_after": self.execute_after,
+            "execute_before": self.execute_before,
+            "calls": [
+                {
+                    "to": call.to_addr,
+                    "selector": call.selector,
+                    "calldata": call.calldata,
+                }
+                for call in self.calls
+            ],
+        }
```

Comment:
Spent another 4 hours trying to fix that. 

And then I remembered: I avoided using serialiser here initially because there is no serialiser that would produce good result. Abi with TypedDict would do work actually, but it is unclear how to create new one. I struggle to find specs, I don't quite undestand which Abi version to extend, and honestly not sure if that is the correct vector for contribution.

This method prepares data to be user with CairoSerialisers, to create CallData out this model.

Existent SchemaSerialisers turn everything into string as they are focused on RPC communication, thus, they don't fit the task because Ints are being turned into Felt strings, and CairoSerialiser deprecates those explicitly in the code. So that is the conflict. 

CairoSerialisers do not allow object as an input. 

I feel like I am stuck with this one. I would suggest to fix it in separate PR by extending Abi models. Could you provide documentation on this ones?

Reply:
Would be great to use TypedDicts for models as much possible.

Reply:
Ok, se we can simply create `OutsideExecutionSchema` and `CallSchema` and use them here. Here is an example of [`TransactionSchema`](https://github.com/software-mansion/starknet.py/blob/45825840ffd013569b82887a9eed89fa26ff41ee/starknet_py/net/schemas/rpc/transactions.py#L121). 

Here's an example code that should work:

```python
class CallSchema(Schema):
    to_addr = Felt(data_key="contract_address", required=True)
    selector = Felt(data_key="entry_point_selector", required=True)
    calldata = fields.List(Felt(), data_key="calldata", required=True)


class OutsideExecutionSchema(Schema):
    caller = Felt(data_key="caller", required=True)
    nonce = Felt(data_key="nonce", required=True)
    execute_after = Uint64(data_key="execute_after", required=True)
    execute_before = Uint64(data_key="execute_before", required=True)
    calls = fields.List(fields.Nested(CallSchema()), data_key="calls", required=True)
```

And then use
```python
return cast(Dict, OutsideExecutionSchema().dump(obj=self))
```

Reply:
Ouch. I'm so sorry you had to write all this. That is exactly the code I wrote myself. And it does not work. The problem is precisely what I've described: 
```
Felt(data_key="contract_address", required=True)
```
produces string of 60-ish symbols. And this in turn breaks serialisation further, (as only short strings are supported by CairoSerialiser)[https://github.com/software-mansion/starknet.py/blob/45825840ffd013569b82887a9eed89fa26ff41ee/starknet_py/serialization/data_serializers/felt_serializer.py#L31].

Reply:
My code works, because it uses int instead of hex int represeantation for Felt, which CairoSerialiser handles perfectly.

Reply:
Hmm 🤔 , but caller is [contract address](https://github.com/starkware-libs/starknet-specs/blob/6d88b7399f56260ece3821c71f9ce53ec55f830b/api/starknet_api_openrpc.json#L1276) which points to [felt](https://github.com/starkware-libs/starknet-specs/blob/6d88b7399f56260ece3821c71f9ce53ec55f830b/api/starknet_api_openrpc.json#L1299). So why is this not valid?

Reply:
When I meant string I meant primitive python type. Serialisers are serialising everything into it's string representation, thus Felt if string of a certain format, and knowledge of the Logical type is encapsulated within the serialiser. 
![image](https://github.com/user-attachments/assets/40ed4aa1-5046-4e33-97c8-a08f10eed0cd)

So we have object, that upon passing through serialiser turns into dict[str, str]. But CairoSerialiser works with primitive python types. So Felt in its logic is just int. And when it is not then there is a branch of logic I've sent you link to.

BTW. Thank you so much for the spec. I'll study it and might come with a refactoring proposal.

Reply:
Ok, got it. Could you please add [this](https://github.com/software-mansion/starknet.py/issues/1537) todo in comment? We may address it in future, as generally it's better to use schema than construct dict by hand.